File: alignment/unmatched_pairs.py
#! /bin/env python

from __future__ import print_function
import sys
import pysam

samfile = pysam.AlignmentFile("/goon2/scratch2/vyp-scratch2/WGS/Hardcastle_October2014/align/data/140724_H02J1_MH-901804_sorted_unique.bam",'rb')

# 'ChrX:139,400,000-139,600,000'
#for reads in samfile.pileup('chrX',139400000,139600000):

for read in samfile.fetch('chrX',139400000,139600000):
    if not read.mate_is_unmapped and samfile.getrname(read.mrnm)!='chrX':
        print(read.mpos)
    else:
        print(read)


reads=[ r for r in samfile.fetch('chrX',139400000,139600000) ]

# percentage of unmapped reads
100 * float(len(filter(lambda x: x.is_unmapped, reads))) / len(reads)

reads=filter(lambda x: not x.is_unmapped, reads)

# reads where the mate maps to another chromosome
for r in filter(lambda x: x.rname!=x.mrnm, reads):
    m=samfile.mate(r)
    print('%s:%d'%(samfile.getrname(r.rname),r.pos,), '%s:%d'%(samfile.getrname(m.rname), m.pos,))


reversed_mate=filter(lambda x: x.mate_is_reverse, reads)


for r in filter(lambda x: not x.mate_is_unmapped and not x.is_unmapped and x.rname==x.mrnm, reads):
    #print(r.is_reverse)
    #print(r.aend-r.pos)
    m=samfile.mate(r)
    #print(m.is_reverse)
    #print(m.aend-m.pos)
    print(m.pos-r.pos)


nice_reads=filter(lambda x: not x.mate_is_unmapped and not x.is_unmapped and x.rname==x.mrnm, reads)
#nice_read_lengths=map(lambda r: samfile.mate(r).pos-r.pos, nice_reads)
nice_read_lengths=map(lambda r: r.mpos-r.pos, nice_reads)

pos_len=filter(lambda x: x>=0, nice_read_lengths)
neg_len=filter(lambda x: x<0, nice_read_lengths)

File: alignment/unmatched_pairs.R
library(Rsamtools)
 
#bamFile <- "C:/Users/Jessica Gardner/Desktop/LP0022129-DNA_A01.bam"
 
# reads in all files in the directory which finish in .bam
print(files <- grep('.bam$',list.files(path="Z:/WGS/data from Niko for Jess 25.08.15/JessicaGardner/",pattern=".*.bam",full.names = TRUE),value = T))
 for (f in files) {
  print(f)
  # skip file if size less than 100 (it is probably empty - only contains header information)
  if (file.info(f)$size<100) next
  # read file
  bams <- scanBam(f)
  # not sure
  b <- bam[[1]]
  # count to which chromosome the mates map
print(sort(table(b$mrnm),decreasing=T))
}
 
bamFile <- "Z:/WGS/bam files/X-LINKED RP/140714_H00TK_PF-19956/140714_H00TK_PF-19956_sorted_unique.bam"
bamFile <- "Z:/WGS/bam files/X-LINKED RP/140724_H02J1_MH-901804/140724_H02J1_MH-901804_sorted_unique.bam"
# "which" is the region you wish to pull out
# "what" are the paramaters which will be present in bam data structure
# a subset may be selected like so:
#what=c("rname", "strand", "pos", "qwidth", "seq", "mrnm", "mpos","strand","mseq")
param <- ScanBamParam(which=RangesList(chrX=IRanges(140410000, 140430000)), what=scanBamWhat()) #,what=scanBamFlag(isPaired = TRUE,isProperPair = TRUE))
bam <- scanBam(bamFile, param=param)
# you need to extract the first item of the list to access the data, I am not sure why yet
b <- bam[[1]]
 # count to which chromosome the mates map and sort
print(sort(table(b$mrnm),decreasing=T))
# mate maps to chromosome 3
i <- which(b$mrnm=='chr3')
b$pos[i]
as.character(b$seq[i])
